### PR TITLE
Fill out advanced configuration format documentation

### LINKED ISSUE #380: sgrep-lint API reference
To make sgrep-lint readily usable by security engineers, the sgrep-lint API should be fully documented.

This should include:

- How to invoke sgrep-lint, as a binary, and as a Docker image
- YAML structure
- Documentation on each key within the YAML schema

### PR DESCRIPTION
**Pull request by mschwager**
Still working towards #380.

### CODE REVIEW

**Comment by mschwager** [comment]
See the resultant markdown here: https://github.com/returntocorp/sgrep/blob/a2bf05b363dd370a5c59e71386eefcad2b030894/docs/config/advanced.md

Still pending in draft status until we figured out `patterns-and` and `patterns-or`.

File: README.md [hunk 1]
```diff
@@ -54,17 +54,19 @@ Start with a simple example:
 
 ```
 $ cat << EOF > test.py
-a = 1
-b = 2
-if a == a:  # oops, supposed to be a == b
-    print('sgrep test')
+def get_node(node_id, nodes):
```
Comment (mschwager) [inline README.md:57]: This example felt less contrived while still staying relatively small and self-contained.

File: docs/config/advanced.md [hunk 2]
```diff
@@ -1,95 +1,275 @@
...
+
+_* Only one of `pattern`, `patterns-and`, or `patterns-or` is required._
+
+**Optional:**
+
+All optional fields must reside underneath a `patterns-and` or `patterns-or` field.
+
+| Field | Type | Description |
+| :--- | :--- | :--- |
+| [`pattern-not`](#pattern-not) | `string` | Logical NOT - memove findings matching this expression. |
```
Comment (ievans) [inline docs/config/advanced.md]:
```suggestion
| [`pattern-not`](#pattern-not) | `string` | Logical NOT - remove findings matching this expression. |
```

**Review by ievans — DISMISSED** [review]
This is great work and a big improvement. 

One comment: currently the primary README refers to the unique sgrep features (metavars + ellipsis) under "configuration operators". To be consistent, I feel they should be grouped with the simple / pattern section? Might be worth referring to that in this document as well.

**Comment by mschwager** [comment]
> This is great work and a big improvement.
> 
> One comment: currently the primary README refers to the unique sgrep features (metavars + ellipsis) under "configuration operators". To be consistent, I feel they should be grouped with the simple / pattern section? Might be worth referring to that in this document as well.

Good catch! Next I'm hoping to unify the `simple` and `advanced` docs into one page. We currently see metavars + ellipsis as sgrep OCaml functionality and pattern operators and configuration files as sgrep-lint Python functionality. But to external users it's all just "configuration", IMO. With that being said, we should still have different names for these concepts (they can't both be "operators"). What do you think of **"Expression features"** for metavars + ellipsis + other OCaml functionality and **"Configuration operators"** for pattern matching + configuration files?

My thoughts on next steps are: unify these two concepts in a single docs page and make the `simple` docs the README (really basic quickstart info) and the `advanced` docs the unified page. I'll patch up the wording in this PR so the two concepts are consistently named then add "Expression features" to the advanced page and unify the pages in a separate PR. Thoughts?

**Comment by ievans** [comment]
@mschwager I would stick to "pattern features" for metavar + ellipsis + other O'Caml functionality since we consistently call them "patterns" everywhere For the configuration one, I don't have a clear sense of what makes the most sense 

I like your plan!

**Comment by aryx** [comment]
I vote for "pattern features/operators" and "boolean/logic operators"
and there will be soon "code equivalence operators" (==> and <==> in a 'equivalences' section).

File: README.md [hunk 3]
```diff
@@ -110,9 +110,9 @@ handles a multitude of input types:
 In the absense of this flag, a default configuration is loaded from `.sgrep.yml`
 or multiple files matching `.sgrep/**/*.yml`.
 
-#### Operators
+#### Pattern Features
 
-Configuration files make use of two primary operators:
+`sgrep` expressions make use of two primary pattern features:
```
Comment (ievans) [inline README.md]:
```suggestion
`sgrep` patterns make use of two primary features:
```

**Review by ievans — APPROVED** [review]

**Pull request merged**